File: game/src/sim/rng.ts
// Seeded RNG with NAMED SUBSTREAMS. No global instance (a determinism
// hazard). The cursor is serialisable so a replay re-seeds exactly.
//
// Generator: sfc32 (Small Fast Counting, 32-bit) — passes PractRand to
// multi-TB, ~2^128 state, far better equidistribution/period than an LCG.
// Seeded via splitmix32 so even nearby seeds decorrelate fully. All math is
// 32-bit (Math.imul / >>>0) so results are bit-identical across machines.

function splitmix32(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x9e3779b9) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 16), 0x21f0aaad) >>> 0;
    t = Math.imul(t ^ (t >>> 15), 0x735a2d97) >>> 0;
    return (t ^ (t >>> 15)) >>> 0;
  };
}

function nameHash(name: string): number {
  let h = 2166136261 >>> 0;
  for (let i = 0; i < name.length; i++) {
    h ^= name.charCodeAt(i);
    h = Math.imul(h, 16777619) >>> 0;
  }
  return h >>> 0;
}

export class Substream {
  private a: number;
  private b: number;
  private c: number;
  private d: number;
  public draws = 0;

  constructor(seed: number) {
    // Mix the seed into four 32-bit words of sfc32 state.
    const sm = splitmix32((seed >>> 0) || 1);
    this.a = sm();
    this.b = sm();
    this.c = sm();
    this.d = sm();
    // Warm up to wash out seeding structure.
    for (let i = 0; i < 16; i++) this.u32();
  }

  private u32(): number {
    // sfc32 core.
    const t = (((this.a + this.b) >>> 0) + this.d) >>> 0;
    this.d = (this.d + 1) >>> 0;
    this.a = (this.b ^ (this.b >>> 9)) >>> 0;
    this.b = (this.c + (this.c << 3)) >>> 0;
    this.c = ((this.c << 21) | (this.c >>> 11)) >>> 0;
    this.c = (this.c + t) >>> 0;
    return t >>> 0;
  }

  /** Uniform float in [0,1). 53-bit mantissa from two 32-bit draws. */
  next(): number {
    this.draws++;
    const hi = this.u32() >>> 5; // 27 bits
    const lo = this.u32() >>> 6; // 26 bits
    return (hi * 67108864 + lo) / 9007199254740992; // / 2^53
  }
  range(min: number, max: number): number {
    return min + this.next() * (max - min);
  }
  int(min: number, max: number): number {
    return Math.floor(this.range(min, max));
  }
  chance(p: number): boolean {
    return this.next() < p;
  }
}

function hash2(a: number, b: number): number {
  let h = (a ^ 0x9e3779b9) >>> 0;
  h = Math.imul(h ^ (h >>> 16), 0x85ebca6b) >>> 0;
  h = (h + (b >>> 0)) >>> 0;
  h = Math.imul(h ^ (h >>> 13), 0xc2b2ae35) >>> 0;
  return (h ^ (h >>> 16)) >>> 0;
}

export class Rng {
  private streams = new Map<string, Substream>();
  constructor(public readonly seed: number) {}
  /** Get (or lazily create) a named substream, deterministically seeded. */
  sub(name: string): Substream {
    let s = this.streams.get(name);
    if (!s) {
      s = new Substream(hash2(this.seed, nameHash(name)));
      this.streams.set(name, s);
    }
    return s;
  }
  /** Serialisable cursor (substream -> draw count) for replay re-seeding. */
  cursor(): Record<string, number> {
    const out: Record<string, number> = {};
    for (const [k, v] of this.streams) out[k] = v.draws;
    return out;
  }
}

File: game/rig-core/src/rng.rs
//! sfc32 + splitmix32 seeding — a BIT-EXACT port of the TS sim/rng.ts so the
//! Rust core and the TS oracle produce identical streams (replay parity).
//! All ops are u32 wrapping, mirroring JS `Math.imul` / `>>> 0`.

use std::collections::HashMap;

fn splitmix32_next(a: &mut u32) -> u32 {
    *a = a.wrapping_add(0x9e37_79b9);
    let mut t = *a;
    t = (t ^ (t >> 16)).wrapping_mul(0x21f0_aaad);
    t = (t ^ (t >> 15)).wrapping_mul(0x735a_2d97);
    t ^ (t >> 15)
}

fn name_hash(name: &str) -> u32 {
    let mut h: u32 = 2166136261;
    for b in name.bytes() {
        h ^= b as u32;
        h = h.wrapping_mul(16777619);
    }
    h
}

fn hash2(a: u32, b: u32) -> u32 {
    let mut h = a ^ 0x9e37_79b9;
    h = (h ^ (h >> 16)).wrapping_mul(0x85eb_ca6b);
    h = h.wrapping_add(b);
    h = (h ^ (h >> 13)).wrapping_mul(0xc2b2_ae35);
    h ^ (h >> 16)
}

#[derive(Clone)]
pub struct Substream {
    a: u32,
    b: u32,
    c: u32,
    d: u32,
    pub draws: u64,
}

impl Substream {
    pub fn new(seed: u32) -> Self {
        let mut sm = if seed == 0 { 1 } else { seed };
        let a = splitmix32_next(&mut sm);
        let b = splitmix32_next(&mut sm);
        let c = splitmix32_next(&mut sm);
        let d = splitmix32_next(&mut sm);
        let mut s = Self { a, b, c, d, draws: 0 };
        for _ in 0..16 {
            s.u32();
        }
        s
    }

    fn u32(&mut self) -> u32 {
        let t = self.a.wrapping_add(self.b).wrapping_add(self.d);
        self.d = self.d.wrapping_add(1);
        self.a = self.b ^ (self.b >> 9);
        self.b = self.c.wrapping_add(self.c << 3);
        self.c = (self.c << 21) | (self.c >> 11);
        self.c = self.c.wrapping_add(t);
        t
    }

    /// Uniform [0,1) — identical f64 to the TS impl (53-bit, two u32 draws).
    pub fn next(&mut self) -> f64 {
        self.draws += 1;
        let hi = (self.u32() >> 5) as f64; // 27 bits
        let lo = (self.u32() >> 6) as f64; // 26 bits
        (hi * 67108864.0 + lo) / 9007199254740992.0
    }
    pub fn range(&mut self, min: f64, max: f64) -> f64 {
        min + self.next() * (max - min)
    }
    pub fn int(&mut self, min: i64, max: i64) -> i64 {
        (self.range(min as f64, max as f64)).floor() as i64
    }
    pub fn chance(&mut self, p: f64) -> bool {
        self.next() < p
    }
}

pub struct Rng {
    seed: u32,
    streams: HashMap<String, Substream>,
}

impl Rng {
    pub fn new(seed: u32) -> Self {
        Self {
            seed,
            streams: HashMap::new(),
        }
    }
    pub fn sub(&mut self, name: &str) -> &mut Substream {
        let seed = self.seed;
        self.streams
            .entry(name.to_string())
            .or_insert_with(|| Substream::new(hash2(seed, name_hash(name))))
    }
    pub fn cursor(&self) -> Vec<(String, u64)> {
        let mut v: Vec<_> = self
            .streams
            .iter()
            .map(|(k, s)| (k.clone(), s.draws))
            .collect();
        v.sort();
        v
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matches_ts_oracle_known_answers() {
        // Captured from the TS impl: new Rng(12345).sub('contest').next() x5
        let expected = [
            0.96806102826403,
            0.14561917944373737,
            0.9415351546655758,
            0.03234286412563092,
            0.26646868810857627,
        ];
        let mut r = Rng::new(12345);
        let s = r.sub("contest");
        for e in expected {
            let v = s.next();
            assert!((v - e).abs() < 1e-15, "got {v}, want {e}");
        }
    }

    #[test]
    fn uniform_mean() {
        let mut r = Rng::new(7);
        let s = r.sub("aiJitter");
        let mut sum = 0.0;
        for _ in 0..100_000 {
            sum += s.next();
        }
        let mean = sum / 100_000.0;
        assert!((mean - 0.498694).abs() < 1e-6, "mean {mean}");
    }

    #[test]
    fn substreams_decorrelate() {
        let mut r = Rng::new(99);
        let u: Vec<f64> = (0..1000).map(|_| r.sub("a").next()).collect();
        let v: Vec<f64> = (0..1000).map(|_| r.sub("b").next()).collect();
        let same = u.iter().zip(&v).filter(|(x, y)| (*x - *y).abs() < 1e-12).count();
        assert_eq!(same, 0);
    }
}
